### ISSUE TITLE
Unit Testing Lookup Rules

### ISSUE THREAD

**Issue by fujiiface** [MarimerLLC/cslaforum#406]
Hi all,

I have implemented a Lookup rule based on the `RuleTutorial` solution's `2.08 LookupRule` project and wanted to unit test it.  Within the `RuleTutorial` solution there is a unit test project for lookup rules which I have mostly followed.  I am a little confused though about the actual lookup part.

The `LookupRule` tutorial has a `LookupCustomerCommand` object which simulates a customer lookup when the `CustomerId` property is set.  The `LookupCustomerTest` also uses this command which simulates a Customer lookup.

**My question:** Do I need to also mock my equivalent `LookupCustomerCommand` object?  Currently, my unit test tries to use my lookup command object which tries to do a real lookup.  This works within my application's context but if I try to do the lookup in the unit test I get a database name not found in config error.

**Comment by ajj7060** [comment]
Yes, you'll have to create an interface which represents the command and mock the interface for unit testing your rule.  Unless you do want to setup the database and actually have the rule hit the database, but IMO that's more an integration test, and tests that hit the DB are much harder to maintain.

You could mock the data access which the LookupCustomerCommand uses as well, and then your rule test would have to mock whatever the command needs to function properly.  But that ends up tying your rule test more to the actual implementation of the command, which can be a bit unwieldy too.
